xcube/util/jsonschema.py

```diff
@@ -175,54 +206,91 @@ def to_dict(self) -> Dict[str, Any]:
         except jsonschema.ValidationError:
             return False
 
-    @staticmethod
-    def new_datetime_range() -> 'JsonArraySchema':
-        """Return a schema for a two-element array of dates or date-times"""
-        return JsonArraySchema(items=[
-                    JsonDatetimeSchema(any_of=[
-                        JsonDatetimeSchema(format='date-time', nullable=True),
-                        JsonDatetimeSchema(format='date', nullable=True),
-                    ]),
-                    JsonDatetimeSchema(any_of=[
-                        JsonDatetimeSchema(format='date-time', nullable=True),
-                        JsonDatetimeSchema(format='date', nullable=True),
-                    ], nullable=True)])
+
+class JsonDateSchema(JsonStringSchema, JsonDateAndTimeSchemaBase):
+    """
+    JSON schema for date instances.
+
+    :param min_date: optional minimum date.
+    :param max_date: optional maximum date.
+    """
+
+    # noinspection PyShadowingBuiltins
+    def __init__(self,
+                 min_date: str = None,
+                 max_date: str = None,
+                 **kwargs):
+        super().__init__(**kwargs, format='date')
+        self._validate_value(min_date, 'min_date', 'date')
+        self._validate_value(max_date, 'max_date', 'date')
+        self.min_date = min_date
+        self.max_date = max_date
+
+    def to_dict(self) -> Dict[str, Any]:
+        d = super().to_dict()
+        if self.min_date is not None:
+            d.update(minDate=self.min_date)
+        if self.max_date is not None:
+            d.update(maxDate=self.max_date)
+        return d
+
+    @classmethod
+    def new_range(cls,
+                  min_date: str = None,
+                  max_date: str = None,
+                  nullable: bool = False) -> 'JsonArraySchema':
+        """
+        Return a schema for a date range.
+        :param min_date: optional minimum date.
+        :param max_date: optional maximum date.
+        :param nullable: whether the whole range as well as individual start and end may be None.
+        :return: a JsonArraySchema with two items.
+        """
+        return JsonArraySchema(
+            items=[
+                JsonDateSchema(min_date=min_date, max_date=max_date, nullable=nullable),
+                JsonDateSchema(min_date=min_date, max_date=max_date, nullable=nullable),
+            ],
+            nullable=nullable)
+
+
+class JsonDatetimeSchema(JsonStringSchema, JsonDateAndTimeSchemaBase):
+    """
+    JSON schema for date-time instances.
+
+    :param min_datetime: optional minimum date-time.
+    :param max_datetime: optional maximum date-time.
+    """
+
+    # noinspection PyShadowingBuiltins
+    def __init__(self,
+                 min_datetime: str = None,
+                 max_datetime: str = None,
+                 **kwargs):
+        super().__init__(**kwargs, format='date-time')
+        JsonDateSchema._validate_value(min_datetime, 'min_datetime', 'date-time')
```

Comment:
In Json DateSchema, you specify this as self._validate_value(...) . The same function is used, but I'd prefer to see it called the same way.

Reply:
It will in any case call through to `JsonDateAndTimeSchemaBase`, but I agree that `JsonDatetimeSchema` going via `JsonDateSchema` to call `_validate_value` is rather awkward. Either `self._validate_value` or `JsonDatetimeSchema._validate_value` would be preferable (personally I prefer the latter since it makes it clear that a class method is being called).

Reply:
Happened by accident.